src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/QueryOrRequestParameterValueResolver.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\Controller\ArgumentResolver;
+
+use Symfony\Component\HttpFoundation\Exception\BadRequestException;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Attribute\QueryParameter;
+use Symfony\Component\HttpKernel\Attribute\RequestParameter;
+use Symfony\Component\HttpKernel\Controller\ValueResolverInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ArgumentMetadata;
+
+final class QueryOrRequestParameterValueResolver implements ValueResolverInterface
+{
+    /**
+     * @return array<mixed>
+     */
+    public function resolve(Request $request, ArgumentMetadata $argument): array
+    {
+        $attribute = $argument->getAttributesOfType(QueryParameter::class)[0] ?? $argument->getAttributesOfType(RequestParameter::class)[0] ?? null;
+        if (null === $attribute) {
+            return [];
+        }
+
+        $inputBag = $attribute instanceof QueryParameter ? $request->query : $request->request;
+
+        $name = $attribute->name ?? $argument->getName();
+        if (false === $inputBag->has($name)) {
+            if ($argument->isNullable() || $argument->hasDefaultValue()) {
+                return [];
+            }
+
+            throw new BadRequestException(sprintf('Missing "%s" parameter "%s".', $attribute instanceof QueryParameter ? 'query' : 'request', $name));
+        }
+
+        if ($argument->isVariadic()) {
+            return $inputBag->all($name);
+        }
+
+        if ('array' === $argument->getType()) {
+            return [$inputBag->all($name)];
+        }
+
+        $value = $inputBag->get($name);
+
+        if ('string' === $argument->getType()) {
+            return [$value];
+        }
+
+        if ('int' === $argument->getType()) {
+            if (!is_numeric($value)) {
+                throw new BadRequestException(sprintf('"%s" parameter "%s" must be an integer.', $attribute instanceof QueryParameter ? 'Query' : 'Request', $name));
```

Comment:
Shouldn't this be a 404 instead?

Reply:
It seems that `NotFoundHttpException` is the recommended way in a ValueResolver. Fixed now.